File: Foundations/Merge-Sort.py
#!/usr/bin/python
"""Programa que ordena una secuencia usando el algoritmo Merge-Sort."""

import random


def Merge(A, p, q, r):
    """Combina dos subsecuencias ordenadas dentro de un arreglo."""
    n1 = q - p + 1
    n2 = r - q

    L = []
    R = []
    for i in range(n1):
        L.append(A[p + i])
    for j in range(n2):
        R.append(A[q + j + 1])
    L.append(9999999)
    R.append(9999999)

    i = 0
    j = 0
    for k in range(p, r + 1):
        if L[i] <= R[j]:
            A[k] = L[i]
            i += 1
        else:
            A[k] = R[j]
            j += 1


def MergeSortR(A, p, r):
    """Ordena la sublista A[p..r] por medio del algoritmo Merge-Sort."""
    if p < r:
        q = (p + r) / 2
        MergeSortR(A, p, q)
        MergeSortR(A, q + 1, r)
        Merge(A, p, q, r)


def MergeSort(A):
    """Ordena la lista A por medio del algoritmo MergeSort."""
    MergeSortR(A, 0, len(A) - 1)


A = random.sample(range(20), 20)
print(A)
MergeSort(A)
print(A)

File: Foundations/Merge-Sort.c
#include <stdio.h>
#define INF 999999

void merge(int A[], int p, int q, int r);
void merge_sort_r(int A[], int p, int r);
void merge_sort(int A[], int n);
void imprimir(int A[], int n);

int main(int argc, char const *argv[]) {

    int A[] = {4, 2, 10, 5, 3, 6, 9, 7, 8, 1};
    imprimir(A, 10);

    merge_sort(A, 10);
    imprimir(A, 10);

    return 0;
}

void merge(int A[], int p, int q, int r)
{
    int n1 = q - p + 1, n2 = r - q;
    int L[n1 + 1], R[n2 + 1];

    int i, j;

    for (i = 0; i < n1; i++) L[i] = A[p + i];
    for (j = 0; j < n2; j++) R[j] = A[q + j + 1];
    L[n1] = INF; R[n2] = INF;

    i = 0; j = 0;
    for (int k = p; k <= r; k++)
    {
        if (L[i] <= R[j])
        {
            A[k] = L[i];
            i += 1;
        }
        else
        {
            A[k] = R[j];
            j += 1;
        }
    }
}

void merge_sort_r(int A[], int p, int r)
{
    int q;

    if (p < r)
    {
        q = (p + r) / 2;
        merge_sort_r(A, p, q);
        merge_sort_r(A, q + 1, r);
        merge(A, p, q, r);
    }
}

void merge_sort(int A[], int n)
{
    merge_sort_r(A, 0, n - 1);
}

void imprimir(int A[], int n)
{
    for (int i = 0; i < n; i++)
        printf("%d ", A[i]);
    printf("\n");
}
